pkg/objstore/cos/cos.go

```diff
@@ -0,0 +1,391 @@
+package cos
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"net/http"
+	"os"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/go-kit/kit/log"
+	"github.com/improbable-eng/thanos/pkg/objstore"
+	"github.com/improbable-eng/thanos/pkg/runutil"
+	"github.com/mozillazg/go-cos"
+	"github.com/pkg/errors"
+	yaml "gopkg.in/yaml.v2"
+)
+
+// DirDelim is the delimiter used to model a directory structure in an object store bucket.
+const dirDelim = "/"
+
+// Bucket implements the store.Bucket interface against cos-compatible(Tencent Object Storage) APIs.
+type Bucket struct {
+	logger log.Logger
+	client *cos.Client
+	name   string
+}
+
+// Config encapsulates the necessary config values to instantiate an cos client.
+type Config struct {
+	Bucket    string `yaml:"bucket"`
+	Region    string `yaml:"region"`
+	AppId     string `yaml:"app_id"`
+	SecretKey string `yaml:"secret_key"`
+	SecretId  string `yaml:"secret_id"`
+}
+
+// Validate checks to see if mandatory cos config options are set.
+func (conf *Config) validate() error {
+	if conf.Bucket == "" ||
+		conf.AppId == "" ||
+		conf.Region == "" ||
+		conf.SecretId == "" ||
+		conf.SecretKey == "" {
+		return errors.New("insufficient cos configuration information")
+	}
+	return nil
+}
+
+func NewBucket(logger log.Logger, conf []byte, component string) (*Bucket, error) {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	var config Config
+	if err := yaml.Unmarshal(conf, &config); err != nil {
+		return nil, errors.Wrap(err, "parsing cos configuration")
+	}
+	if err := config.validate(); err != nil {
+		return nil, errors.Wrap(err, "validate cos configuration")
+	}
+
+	bucketUrl := cos.NewBucketURL(config.Bucket, config.AppId, config.Region, true)
+
+	b, err := cos.NewBaseURL(bucketUrl.String())
+	if err != nil {
+		return nil, errors.Wrap(err, "initialize cos base url")
+	}
+
+	client := cos.NewClient(b, &http.Client{
+		Transport: &cos.AuthorizationTransport{
+			SecretID:  config.SecretId,
+			SecretKey: config.SecretKey,
+		},
+	})
+
+	bkt := &Bucket{
+		logger: logger,
+		client: client,
+		name:   config.Bucket,
+	}
+	return bkt, nil
+}
+
+// Name returns the bucket name for s3.
+func (b *Bucket) Name() string {
+	return b.name
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	reader, length, err := objectLength(r)
+	if err != nil {
+		return errors.Wrap(err, "length of cos object")
+	}
+	opt := &cos.ObjectPutOptions{
+		ObjectPutHeaderOptions: &cos.ObjectPutHeaderOptions{
+			ContentLength: length,
+		},
+	}
+
+	// The reason wrapping the reader as a nop closer is to avoid to close it after uploading in COS client.
```

Comment:
Not needed at this moment, because I have committed a PR to the upstream [PR](https://github.com/mozillazg/go-cos/pull/7). Before the PR, it would hit the `os.ErrClosed` in Thanos, because the `net/http` package will close the request body automatically for safety after requesting.